File: prime.rb
#!/usr/bin/env ruby
# -*- coding: utf-8 -*-


def primes(n)
    # sieve of Eratosthenes
    f = Array.new(n+1) { |i| i % 2}
    i = 3
    nn = Math.sqrt(n)

    while i <= nn do
        if f[i] == 1
            j = i * i
            while j <= n do
                f[j] = 0
                j += 2 * i  #i.e. j=i(i+2) not j += i, since 2|i(i+1)
            end
        end
        i += 2
    end
 
    Array.new( (n -1) / 2 ) { |i| 3 + 2 * i }.each do |x|
        if f[x] == 1
            $P << x
        end
    end
 
end
 
t1 = Time.now

$P = [2]

n = 10000000
primes(n)

t2 = Time.now
delta = t2 - t1
puts delta

File: prime.jl
#!/usr/bin/env julia

function primes(n)
    # sieve of Eratosthenes
    f = zeros(n)
    for i = 2:2:n
        f[i] = 1
    end

    i = 3
    nn = sqrt(n)
    while i <= nn
        if f[i] == 0
            j = i * i
            while j <= n
                f[j] = 1
                j += 2 * i  #i.e. j=i(i+2) not j += i, since 2|i(i+1)
            end
        end
        i += 2
    end
    
    P = [2]
    for x in 3:2:n
        if f[x] == 0
            append!(P, x)
        end
    end
    return P
 
end
 

n = 10000000
@time P= primes(n)

